README.md

```diff
@@ -197,6 +197,60 @@ Use `settings.jsdoc.tagNamePreference` to configure a preferred alias name for a
 }
 ```
 
+One may also use an object with a `message` and `replacement`, with `{{tagName}}` and `{{preferredType}}` (or its alias `{{replacement}}`) as template variables to be
```

Comment:
This template replacement thing looks like an excessive feature for me. In terms of UX it easier to understand and maintain a setting like this
```
"message": "@extends is preferred over @arguments."
``` 
than this
```
"message": "@{{replacement}} is preferred over @{{tagName}}."`
```

There might be rare cases where this feature helps, but the benefits cannot justify the complexity it brings.

Reply:
eslint already has templating in place, so there is not even a need for `.replace()` calls here--just pass the variables in. Also, I've already implemented this approach with `preferredTypes`.

If I am to undo then I think it ought to be a breaking change to remove from `preferredTypes` as well.

I agree though that in this case it doesn't add much, but I thought it should mirror behavior in `preferredTypes`. In `preferredTypes`, I'll admit that `replacement` and `badType` are not super useful there either (though they can keep consistency in messages across different replacements), but I figured it wasn't much to make those variables available since I already needed substitution variables for `tagName` and `tagValue` (both of which cannot be known ahead of time in the case of `preferredTypes` and which were therefore I think justified).

So, if I am to remove template variables in both, I'd be ok with removing the static ones (though requiring a breaking change), but not the dynamic ones.

Reply:
What are static ones and dynamics ones?

Reply:
By "static", I just mean that they can be replaced by the user hard-coding the value:

- `check-types` / `settings.jsdoc.preferredTypes`: **Static**: `badType`, `preferredType`/`replacement`; **Dynamic**: `tagName`, `tagValue`
- `check-tag-names`/`settings.jsdoc.tagNamePreference`:  **All are static**:`preferredTagName`/`replacement`, `tagName`

(`tagName` is static in one and dynamic in the other since when checking types, the tag name may vary, but not when tags are what is being checked.)

Reply:
That's good.

Reply:
So you mean it is good to remove the static ones and keep the dynamic ones?

Reply:
Yes

Reply:
You know, on further reflection though, I'm not sure removing the static ones is such a good idea. Projects may sometimes change the type names they want to use, e.g., "Any", "*", "any", and with those "static" types, they avoid maintenance problems if there is a discrepancy between the `replacement` value and the replacement (or target) type used in the message.

For example, with this code:

```js
/**
  * @param {abc} foo
  */
 function qux(foo) {
 }
```

to get the following error (where one wanted "Abc" to be used in place of "abc"):

> Messed up JSDoc @param "foo" type "abc"; prefer: "Abc".

...with the following, if you wanted to replace `Abc` with some other preferred type (or if you were changing the type to target), you'd have to remember to make the switch in two places:

```js
          preferredTypes: {
            abc: {
              message: 'Messed up JSDoc @{{tagName}}{{tagValue}} type "abc"; prefer: "Abc".',
              replacement: 'Abc'
            }
          }
```

whereas the following doesn't suffer from that problem:

```js
          preferredTypes: {
            abc: {
              message: 'Messed up JSDoc @{{tagName}}{{tagValue}} type "{{badType}}"; prefer: "{{replacement}}".',
              replacement: 'Abc'
            }
          }
```

And passing in the "static" variables hardly takes up any coding lines at all (just passing the variable names in), and as mentioned eslint handles the replacements for us.

Maybe @gajus has an opinion?

Reply:
The above comment seemed to be about `preferredTypes`, does it apply to `tagNamePreference`?

Reply:
Yes, it's the same principle... The "static" properties still allow avoiding of duplication which could be a problem if one changes the tag target (or replacement).

```js
      settings: {
        jsdoc: {
          tagNamePreference: {
            todo: {
              message: 'Please use {{preferredTagName}} instead of {{tagName}}',
              replacement: 'x-todo'
            }
          }
        }
      }
```

Reply:
I still consider this a feature bloat that will be useful for very few people. Adding such a feature will negatively affect users because they have to read longer docs. And the developers have to maintain them.

Reply:
And I don't think it is much extra to "maintain" or read to provide that ability. So let's wait on @gajus . While we need to make code maintainable, I think we ought to err more on the side of being flexible for users.

Reply:
> I still consider this a feature bloat that will be useful for very few people. Adding such a feature will negatively affect users because they have to read longer docs. And the developers have to maintain them.

I agree. Sounds like nuisance to everyone for the benefit of very few.